ibft/roundtimer/timer.go

```diff
@@ -9,78 +9,100 @@ import (
 type RoundTimer struct {
 	timer   *time.Timer
 	cancelC chan bool
+	lapsedC chan bool
 	resC    chan bool
 
-	stopped     bool
-	stoppedLock sync.Mutex
+	stopped  bool
+	syncLock sync.Mutex
 }
 
 // New returns a new instance of RoundTimer
 func New() *RoundTimer {
-	return &RoundTimer{
-		timer:       nil,
-		cancelC:     make(chan bool),
-		resC:        make(chan bool),
-		stopped:     true,
-		stoppedLock: sync.Mutex{},
+	ret := &RoundTimer{
+		timer:    nil,
+		cancelC:  make(chan bool),
+		lapsedC:  make(chan bool),
+		stopped:  true,
+		syncLock: sync.Mutex{},
 	}
+	go ret.eventLoop()
+	return ret
 }
 
 // ResultChan returns the result chan
 // true if the timer lapsed or false if it was stopped
 func (t *RoundTimer) ResultChan() chan bool {
+	t.syncLock.Lock()
+	defer t.syncLock.Unlock()
+
+	if t.resC == nil {
+		t.resC = make(chan bool)
+	}
 	return t.resC
 }
 
+func (t *RoundTimer) fireChannelEvent(value bool) {
+	t.syncLock.Lock()
+	defer t.syncLock.Unlock()
+
+	if t.resC != nil {
+		t.resC <- value
+	}
+}
+
 // Reset will return a channel that sends true if the timer lapsed or false if it was cancelled
 // If Start is called more than once, the first timer and chan are returned and used
 func (t *RoundTimer) Reset(d time.Duration) {
-	t.stoppedLock.Lock()
-	defer t.stoppedLock.Unlock()
+	t.syncLock.Lock()
+	defer t.syncLock.Unlock()
+
+	t.stopped = false
 
 	if t.timer != nil {
 		t.timer.Stop()
 		t.timer.Reset(d)
 	} else {
-		t.timer = time.NewTimer(d)
+		t.timer = time.AfterFunc(d, func() {
```

Comment:
Using `timer.AfterFunc` could lead to leaks as it is not GC'ed:
https://www.fatalerrors.org/a/use-with-caution-time.after-can-cause-memory-leak-golang.html (this suggest to use time ticker instead) 
https://medium.com/@oboturov/golang-time-after-is-not-garbage-collected-4cbc94740082

Reply:
@amir-blox it seems AfterFunc does get GC'd .. only the After function isn't